// @gooddata/sdk-backend-tiger/esm/backend/workspace/permissions/index.js
import { __awaiter, __generator } from "tslib";
var TigerWorkspacePermissionsFactory = /** @class */ (function () {
    function TigerWorkspacePermissionsFactory(authCall, workspace) {
        this.authCall = authCall;
        this.workspace = workspace;
    }
    TigerWorkspacePermissionsFactory.prototype.getPermissionsForCurrentUser = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.authCall(function () { return __awaiter(_this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                return [2 /*return*/, {
                                        canAccessWorkbench: true,
                                        canCreateReport: true,
                                        canCreateVisualization: true,
                                        canExecuteRaw: true,
                                        canExportReport: true,
                                        canManageAnalyticalDashboard: true,
                                        canUploadNonProductionCSV: true,
                                        canManageProject: true,
                                        canCreateAnalyticalDashboard: true,
                                        canInitData: true,
                                        canManageMetric: false,
                                        canManageReport: true,
                                        canCreateScheduledMail: true,
                                        canListUsersInProject: true,
                                        canManageDomain: true,
                                        canInviteUserToProject: true,
                                        canRefreshData: true,
                                        canManageACL: true,
                                    }];
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return TigerWorkspacePermissionsFactory;
}());
export { TigerWorkspacePermissionsFactory };
