// @gooddata/sdk-ui-kit/esm/Dialog/ShareDialog/shareDialogMappers.js
import { __read, __spread } from "tslib";
// (C) 2021 GoodData Corporation
import { areObjRefsEqual } from "@gooddata/sdk-model";
import { notInArrayFilter, GranteeGroupAll, InactiveOwner } from "./ShareDialogBase/utils";
/**
 * @internal
 */
export var mapUserFullName = function (user) {
    if (user.fullName) {
        return user.fullName;
    }
    return user.firstName + " " + user.lastName;
};
/**
 * @internal
 */
export var mapOwnerToGrantee = function (user, currentUserRef) {
    return {
        type: "user",
        id: user.ref,
        name: mapUserFullName(user),
        email: user.email,
        isOwner: true,
        isCurrentUser: areObjRefsEqual(user.ref, currentUserRef),
    };
};
export var mapUserToInactiveGrantee = function () {
    return InactiveOwner;
};
/**
 * @internal
 */
export var mapShareStatusToGroupAll = function (shareStatus) {
    if (shareStatus === "public") {
        return GranteeGroupAll;
    }
};
/**
 * @internal
 */
export var mapGranteesToShareStatus = function (grantees, granteesToAdd, granteesToDelete) {
    var omitDeleted = notInArrayFilter(grantees, granteesToDelete);
    var withAdded = __spread(omitDeleted, granteesToAdd);
    if (withAdded.some(function (g) { return areObjRefsEqual(g.id, GranteeGroupAll.id); })) {
        return "public";
    }
    return "private";
};
