// @gooddata/sdk-ui-dashboard/esm/model/queries/widgets.js
// (C) 2021 GoodData Corporation
/**
 * Creates action thought which you can query dashboard component for filters that should be used by a given widget.
 *
 * @param widgetRef - reference to insight widget
 * @param widgetFilterOverrides - optionally specify filters to be applied on top of the dashboard and insight filters
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing
 * @alpha
 */
export function queryWidgetFilters(widgetRef, widgetFilterOverrides, correlationId) {
    return {
        type: "GDC.DASH/QUERY.WIDGET.FILTERS",
        correlationId: correlationId,
        payload: {
            widgetRef: widgetRef,
            widgetFilterOverrides: widgetFilterOverrides,
        },
    };
}
/**
 *  Creates action thought which you can query dashboard component for broken alert filters.
 *
 * @param widgetRef - reference to insight kpi widget
 * @param correlationId - optionally specify correlation id to use for this command.
 * @returns
 *
 * @alpha
 */
export function queryWidgetBrokenAlerts(widgetRef, correlationId) {
    return {
        type: "GDC.DASH/QUERY.WIDGET.BROKEN_ALERTS",
        correlationId: correlationId,
        payload: {
            widgetRef: widgetRef,
        },
    };
}
