// @gooddata/sdk-ui-dashboard/esm/presentation/scheduledEmail/DefaultScheduledEmailDialog/useScheduledEmail.js
// (C) 2019-2021 GoodData Corporation
import { useCallback } from "react";
import { useDashboardSelector, selectDashboardTitle, selectDashboardUriRef, selectUser, selectLocale, selectFilterContextFilters, selectDateFormat, selectEnableKPIDashboardScheduleRecipients, selectCanListUsersInWorkspace, selectEnableKPIDashboardSchedule, } from "../../../model";
import { useCreateScheduledEmail } from "./useCreateScheduledEmail";
import { invariant } from "ts-invariant";
export var useScheduledEmail = function (props) {
    var onSubmit = props.onSubmit, onSubmitSuccess = props.onSubmitSuccess, onSubmitError = props.onSubmitError;
    // Bear model expects that all refs are sanitized to uriRefs.
    var dashboardUriRef = useDashboardSelector(selectDashboardUriRef);
    // if this bombs then the controller code is bugged because it should not even allow to get
    // to this point for dashboards that are not persisted. scheduling is not possible for such
    // dashboards and so the respective menus to trigger the scheduling must not be present
    invariant(dashboardUriRef, "attempting to schedule email for unsaved dashboard");
    var dashboardTitle = useDashboardSelector(selectDashboardTitle);
    var currentUser = useDashboardSelector(selectUser);
    var locale = useDashboardSelector(selectLocale);
    var filters = useDashboardSelector(selectFilterContextFilters);
    var dateFormat = useDashboardSelector(selectDateFormat);
    var enableKPIDashboardScheduleRecipients = useDashboardSelector(selectEnableKPIDashboardScheduleRecipients);
    var canListUsersInWorkspace = useDashboardSelector(selectCanListUsersInWorkspace);
    var enableKPIDashboardSchedule = useDashboardSelector(selectEnableKPIDashboardSchedule);
    var scheduledEmailCreator = useCreateScheduledEmail({
        onSuccess: onSubmitSuccess,
        onError: onSubmitError,
        onBeforeRun: onSubmit,
    });
    var handleCreateScheduledEmail = useCallback(function (scheduledEmail, customFilters) {
        scheduledEmailCreator.create(scheduledEmail, customFilters !== null && customFilters !== void 0 ? customFilters : filters);
    }, [filters]);
    var scheduledEmailCreationStatus = scheduledEmailCreator.creationStatus;
    return {
        dashboardRef: dashboardUriRef,
        dashboardTitle: dashboardTitle,
        canListUsersInWorkspace: canListUsersInWorkspace,
        enableKPIDashboardSchedule: enableKPIDashboardSchedule,
        enableKPIDashboardScheduleRecipients: enableKPIDashboardScheduleRecipients,
        dateFormat: dateFormat,
        currentUser: currentUser,
        locale: locale,
        handleCreateScheduledEmail: handleCreateScheduledEmail,
        scheduledEmailCreationStatus: scheduledEmailCreationStatus,
    };
};
