// @gooddata/sdk-ui-dashboard/esm/presentation/dashboard/DashboardHeader/ShareDialogDashboardHeader.js
// (C) 2020 GoodData Corporation
import React, { useCallback, useMemo } from "react";
import { useToastMessage } from "@gooddata/sdk-ui-kit";
import { useDashboardSelector, selectIsShareDialogOpen, useDashboardDispatch, uiActions, selectPersistedDashboard, selectCurrentUserRef, useDashboardCommandProcessing, changeSharing, } from "../../../model";
import { ShareDialog } from "../../shareDialog";
var useShareDialogDashboardHeader = function () {
    var dispatch = useDashboardDispatch();
    var _a = useToastMessage(), addSuccess = _a.addSuccess, addError = _a.addError;
    var isShareDialogOpen = useDashboardSelector(selectIsShareDialogOpen);
    var persistedDashboard = useDashboardSelector(selectPersistedDashboard);
    var currentUserRef = useDashboardSelector(selectCurrentUserRef);
    var closeShareDialog = useMemo(function () {
        return function () { return dispatch(uiActions.closeShareDialog()); };
    }, [dispatch, uiActions]);
    var runChangeSharing = useDashboardCommandProcessing({
        commandCreator: changeSharing,
        successEvent: "GDC.DASH/EVT.SHARING.CHANGED",
        errorEvent: "GDC.DASH/EVT.COMMAND.FAILED",
        onSuccess: function () {
            addSuccess({ id: "messages.sharingChangedSuccess" });
        },
        onError: function () {
            addError({ id: "messages.sharingChangedError.general" });
        },
    }).run;
    var onCloseShareDialog = useCallback(function () {
        closeShareDialog();
    }, [closeShareDialog]);
    var onApplyShareDialog = useCallback(function (payload) {
        closeShareDialog();
        runChangeSharing(payload);
    }, [closeShareDialog, runChangeSharing]);
    return {
        isShareDialogOpen: isShareDialogOpen,
        persistedDashboard: persistedDashboard,
        currentUserRef: currentUserRef,
        onCloseShareDialog: onCloseShareDialog,
        onApplyShareDialog: onApplyShareDialog,
    };
};
/**
 * @internal
 */
export var ShareDialogDashboardHeader = function () {
    var _a = useShareDialogDashboardHeader(), isShareDialogOpen = _a.isShareDialogOpen, persistedDashboard = _a.persistedDashboard, currentUserRef = _a.currentUserRef, onCloseShareDialog = _a.onCloseShareDialog, onApplyShareDialog = _a.onApplyShareDialog;
    if (!isShareDialogOpen) {
        return null;
    }
    return (React.createElement(ShareDialog, { isVisible: isShareDialogOpen, currentUserRef: currentUserRef, sharedObject: persistedDashboard, onCancel: onCloseShareDialog, onApply: onApplyShareDialog }));
};
