// @gooddata/sdk-backend-bear/esm/backend/workspace/index.js
// (C) 2019-2021 GoodData Corporation
import { __awaiter, __generator } from "tslib";
import { BearExecution } from "./execution/executionFactory";
import { BearWorkspaceMeasures } from "./measures";
import { BearWorkspaceStyling } from "./styling/styling";
import { BearWorkspaceCatalogFactory } from "./catalog/factory";
import { BearWorkspaceSettings } from "./settings/settings";
import { BearWorkspacePermissionsFactory } from "./permissions/permissions";
import { BearWorkspaceInsights } from "./insights";
import { BearWorkspaceDataSets } from "./datasets";
import { BearWorkspaceDashboards } from "./dashboards";
import { BearWorkspaceUsersQuery } from "./users";
import { BearWorkspaceDateFilterConfigsQuery } from "./dateFilterConfigs";
import { BearWorkspaceAttributes } from "./attributes/index";
import { BearWorkspaceFacts } from "./facts";
import { userLoginMd5FromAuthenticatedPrincipal } from "../../utils/api";
var BearWorkspace = /** @class */ (function () {
    function BearWorkspace(authCall, workspace, descriptor) {
        this.authCall = authCall;
        this.workspace = workspace;
        this.descriptor = descriptor;
    }
    BearWorkspace.prototype.getDescriptor = function () {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var projects, project;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!!this.descriptor) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.authCall(function (sdk, _a) {
                                var getPrincipal = _a.getPrincipal;
                                return __awaiter(_this, void 0, void 0, function () {
                                    var userId;
                                    return __generator(this, function (_b) {
                                        switch (_b.label) {
                                            case 0: return [4 /*yield*/, userLoginMd5FromAuthenticatedPrincipal(getPrincipal)];
                                            case 1:
                                                userId = _b.sent();
                                                // TODO: this is wasteful; we should get single project directly
                                                return [2 /*return*/, sdk.project.getProjects(userId)];
                                        }
                                    });
                                });
                            })];
                    case 1:
                        projects = _c.sent();
                        project = projects.find(function (project) { var _a; return ((_a = project.links) === null || _a === void 0 ? void 0 : _a.self.split("/").pop()) === _this.workspace; });
                        return [2 /*return*/, {
                                id: this.workspace,
                                description: (_a = project === null || project === void 0 ? void 0 : project.meta.summary) !== null && _a !== void 0 ? _a : "",
                                title: (_b = project === null || project === void 0 ? void 0 : project.meta.title) !== null && _b !== void 0 ? _b : "",
                            }];
                    case 2: return [2 /*return*/, this.descriptor];
                }
            });
        });
    };
    BearWorkspace.prototype.getParentWorkspace = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                // Bear has no workspace parenting
                return [2 /*return*/, undefined];
            });
        });
    };
    BearWorkspace.prototype.attributes = function () {
        return new BearWorkspaceAttributes(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.execution = function () {
        return new BearExecution(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.settings = function () {
        return new BearWorkspaceSettings(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.insights = function () {
        return new BearWorkspaceInsights(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.dashboards = function () {
        return new BearWorkspaceDashboards(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.measures = function () {
        return new BearWorkspaceMeasures(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.facts = function () {
        return new BearWorkspaceFacts(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.styling = function () {
        return new BearWorkspaceStyling(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.catalog = function () {
        return new BearWorkspaceCatalogFactory(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.datasets = function () {
        return new BearWorkspaceDataSets(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.permissions = function () {
        return new BearWorkspacePermissionsFactory(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.users = function () {
        return new BearWorkspaceUsersQuery(this.authCall, this.workspace);
    };
    BearWorkspace.prototype.dateFilterConfigs = function () {
        return new BearWorkspaceDateFilterConfigsQuery(this.authCall, this.workspace);
    };
    return BearWorkspace;
}());
export { BearWorkspace };
