// @gooddata/sdk-backend-bear/esm/convertors/fromBackend/UsersConverter.js
import { uriRef } from "@gooddata/sdk-model";
var getUserFullName = function (user) {
    var firstName = user.firstName, lastName = user.lastName;
    if (!firstName && !lastName) {
        return undefined;
    }
    return [firstName, lastName].filter(Boolean).join(" ");
};
export var convertUser = function (user) {
    var email = user.email, login = user.login, firstName = user.firstName, lastName = user.lastName, links = user.links;
    return {
        ref: uriRef(links.self),
        email: email,
        login: login,
        firstName: firstName,
        lastName: lastName,
        fullName: getUserFullName(user),
    };
};
export var convertWorkspaceUser = function (user) {
    var email = user.email, login = user.login, uri = user.uri, firstName = user.firstName, lastName = user.lastName;
    return {
        ref: uriRef(uri),
        email: email,
        login: login,
        uri: uri,
        firstName: firstName !== null && firstName !== void 0 ? firstName : undefined,
        lastName: lastName !== null && lastName !== void 0 ? lastName : undefined,
        fullName: getUserFullName(user),
    };
};
