// @gooddata/sdk-backend-bear/esm/backend/workspace/users/index.js
import { __awaiter, __generator } from "tslib";
import { convertWorkspaceUser } from "../../../convertors/fromBackend/UsersConverter";
var BearWorkspaceUsersQuery = /** @class */ (function () {
    function BearWorkspaceUsersQuery(authCall, workspace) {
        this.authCall = authCall;
        this.workspace = workspace;
        this.options = {};
    }
    BearWorkspaceUsersQuery.prototype.withOptions = function (options) {
        this.options = options;
        return this;
    };
    BearWorkspaceUsersQuery.prototype.queryAll = function () {
        return __awaiter(this, void 0, void 0, function () {
            var search;
            var _this = this;
            return __generator(this, function (_a) {
                search = this.options.search;
                return [2 /*return*/, this.authCall(function (sdk) {
                        return sdk.project
                            .getUserList(_this.workspace, { prefixSearch: search, userState: "ACTIVE" })
                            .then(function (users) { return users.map(convertWorkspaceUser); });
                    })];
            });
        });
    };
    return BearWorkspaceUsersQuery;
}());
export { BearWorkspaceUsersQuery };
