// @gooddata/sdk-ui-pivot/esm/impl/base/agApiWrapper.js
function getHeaderHeight(gridApi) {
    var _a, _b, _c, _d;
    return (_d = (_c = (_b = (_a = gridApi) === null || _a === void 0 ? void 0 : _a.headerRootComp) === null || _b === void 0 ? void 0 : _b.eHeaderContainer) === null || _c === void 0 ? void 0 : _c.clientHeight) !== null && _d !== void 0 ? _d : 0;
}
function getCellElement(gridApi, attributeId, rowIndex) {
    var rowRenderer = gridApi.rowRenderer;
    var rowComp = rowRenderer.rowCompsByIndex[rowIndex];
    return rowComp && rowComp.cellComps[attributeId] ? rowComp.cellComps[attributeId].eGui : null;
}
function addCellClass(gridApi, attributeId, rowIndex, className) {
    var cellElement = getCellElement(gridApi, attributeId, rowIndex);
    if (cellElement !== null) {
        cellElement.classList.add(className);
    }
}
function removeCellClass(gridApi, attributeId, rowIndex, className) {
    var cellElement = getCellElement(gridApi, attributeId, rowIndex);
    if (cellElement !== null) {
        cellElement.classList.remove(className);
    }
}
function getPaginationBottomRowIndex(gridApi) {
    var _a, _b;
    var paginationProxy = gridApi.paginationProxy;
    if (paginationProxy) {
        return (_b = (_a = paginationProxy.bottomRowBounds) === null || _a === void 0 ? void 0 : _a.rowIndex) !== null && _b !== void 0 ? _b : null;
    }
    return null;
}
function getPinnedTopRow(gridApi) {
    var pinnedTopRow = gridApi.rowRenderer.floatingTopRowComps[0];
    return pinnedTopRow ? pinnedTopRow : null;
}
function getPinnedTopRowElement(gridApi) {
    var pinnedTopRow = getPinnedTopRow(gridApi);
    return pinnedTopRow ? pinnedTopRow.bodyContainerComp.eContainer.parentElement.parentElement : null;
}
function addPinnedTopRowClass(gridApi, className) {
    var rowElement = getPinnedTopRowElement(gridApi);
    if (rowElement) {
        rowElement.classList.add(className);
    }
}
function removePinnedTopRowClass(gridApi, className) {
    var rowElement = getPinnedTopRowElement(gridApi);
    if (rowElement) {
        rowElement.classList.remove(className);
    }
}
function setPinnedTopRowStyle(gridApi, propertyName, propertyValue) {
    var rowElement = getPinnedTopRowElement(gridApi);
    if (rowElement) {
        rowElement.style[propertyName] = propertyValue;
    }
}
function getPinnedTopRowCellElementWrapper(gridApi, attributeId) {
    var pinnedTopRow = getPinnedTopRow(gridApi);
    if (!pinnedTopRow) {
        return null;
    }
    var columnIndex = Object.keys(pinnedTopRow.cellComps).find(function (index) {
        return index.slice(0, attributeId.length) === attributeId && pinnedTopRow.cellComps[index] !== null;
    });
    return columnIndex !== undefined && pinnedTopRow.cellComps[columnIndex]
        ? pinnedTopRow.cellComps[columnIndex].eGui
        : null;
}
function getPinnedTopRowCellElement(gridApi, attributeId) {
    var pinnedTopRowCellElementWrapper = getPinnedTopRowCellElementWrapper(gridApi, attributeId);
    return pinnedTopRowCellElementWrapper ? pinnedTopRowCellElementWrapper.querySelector("span") : null;
}
export default {
    getHeaderHeight: getHeaderHeight,
    // cell element
    getCellElement: getCellElement,
    addCellClass: addCellClass,
    removeCellClass: removeCellClass,
    // pinned row element
    getPinnedTopRowElement: getPinnedTopRowElement,
    addPinnedTopRowClass: addPinnedTopRowClass,
    removePinnedTopRowClass: removePinnedTopRowClass,
    setPinnedTopRowStyle: setPinnedTopRowStyle,
    // pinned row cell element
    getPinnedTopRowCellElement: getPinnedTopRowCellElement,
    getPinnedTopRowCellElementWrapper: getPinnedTopRowCellElementWrapper,
    getPaginationBottomRowIndex: getPaginationBottomRowIndex,
};
