// @gooddata/sdk-ui-ext/esm/dataLoaders/InsightDataLoader.js
// (C) 2021 GoodData Corporation
import LRUCache from "lru-cache";
import { objRefToString } from "@gooddata/sdk-model";
import { INSIGHT_CACHE_SIZE } from "./constants";
import { dataLoaderAbstractFactory } from "./DataLoaderAbstractFactory";
var InsightDataLoader = /** @class */ (function () {
    function InsightDataLoader(workspace) {
        this.workspace = workspace;
        this.insightCache = new LRUCache({
            max: INSIGHT_CACHE_SIZE,
        });
    }
    InsightDataLoader.prototype.getInsight = function (backend, ref) {
        var _this = this;
        var cacheKey = objRefToString(ref);
        var insight = this.insightCache.get(cacheKey);
        if (!insight) {
            insight = backend
                .workspace(this.workspace)
                .insights()
                .getInsight(ref)
                .catch(function (error) {
                _this.insightCache.del(cacheKey);
                throw error;
            });
            this.insightCache.set(cacheKey, insight);
        }
        return insight;
    };
    return InsightDataLoader;
}());
/**
 * @internal
 */
export var insightDataLoaderFactory = dataLoaderAbstractFactory(function (workspace) { return new InsightDataLoader(workspace); });
